### PR TITLE
feat(api): add custom user model

### PR DESCRIPTION
**Pull request by divyanshkhetan**

### CODE REVIEW

**Comment by Samy-33** [comment]
@divyanshkhetan @Saksham-1508, please fix the linting errors.

File: api/api/accounts/models.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import datetime
+from django.db import models
+from django.contrib.auth.models import AbstractUser, BaseUserManager, PermissionsMixin
+from django.contrib.auth import get_user_model
+
+# Create your models here.
```
Comment (Samy-33) [inline api/api/accounts/models.py:6]: this comment can be removed.

File: api/api/accounts/models.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    TITLE = (
+        ('Prof', 'Prof'),
+        ('Dr', 'Dr'),
+        ('Mr', 'Mr'),
+        ('Mrs', 'Mrs'),
+        ('Ms', 'Ms'),
+    )
+
+
+class UserManager(BaseUserManager):
```
Comment (Samy-33) [inline api/api/accounts/models.py:44]: Is this a duplicate? We already have `CustomUserManager` in `managers.py`

File: api/api/accounts/models.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        user.is_staff = True
+        user.is_superuser = True
+        user.save(using=self._db)
+        return user
+
+
+class User(AbstractUser, PermissionsMixin):
+    username = None
+    email = models.EmailField(unique=True)
+    gender = models.CharField(max_length=50, choices=Constants.SEX_CHOICES, default="M")
```
Comment (Samy-33) [inline api/api/accounts/models.py:80]: max_length 50 doesn't make sense. Keep it 10.

File: api/api/accounts/models.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        )
+    is_active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=False)
+    is_superuser = models.BooleanField(default=False)
+
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = []
+
+    # user manager
+    objects = UserManager()
```
Comment (Samy-33) [inline api/api/accounts/models.py:96]: Should be `CustomUserModel`

File: api/api/accounts/models.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    def __str__(self):
+        return self.email
+
+    def save(self, *args, **kwargs):
+        """For cpanel."""
+        self.is_active = (self.is_active is True)
+        self.is_staff = (self.is_staff is True)
+        self.is_superuser = (self.is_superuser is True)
```
Comment (Samy-33) [inline api/api/accounts/models.py:105]: Q: What does `(self.is_active is True)` do?

File: api/api/accounts/models.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    hostel_address = models.CharField(verbose_name='Hostel Address', max_length=200)
+
+    # bio
+    bio = models.TextField(max_length=1000, blank=True, null=True)
+
+    # socials
+    linkedin = models.URLField(null=True, blank=True, default="www.linkedin.com")
+    facebook = models.URLField(null=True, blank=True, default="www.facebook.com")
+    instagram = models.CharField(null=True, blank=True, max_length=1000)
+    github = models.URLField(null=True, blank=True, default="www.github.com")
```
Comment (Samy-33) [inline api/api/accounts/models.py:125]: There should be a separate model for these: `StudentSocial`. And default should be null.

File: api/api/accounts/models.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    def __str__(self):
+        return self.user.email
+
+
+class Faculty(models.Model):
+    user = models.OneToOneField(get_user_model(), on_delete=models.CASCADE)
+    title = models.CharField(max_length=20, choices=Constants.TITLE)
+    department = models.CharField(max_length=100, choices=Constants.BRANCH)
+    designation = models.CharField(max_length=100, null=True)
+    # profile_img = models.ImageField(null=True)
```
Comment (Samy-33) [inline api/api/accounts/models.py:136]: Stale comment should be removed.

File: api/api/accounts/models.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    def __str__(self):
+        return self.user.email
+
+
+class Staff(models.Model):
+    user = models.OneToOneField(get_user_model(), on_delete=models.CASCADE)
+    # TODO: choices in department yet to be added
```
Comment (Samy-33) [inline api/api/accounts/models.py:144]: Create an issue and add the issue number in todo like: `TODO(#issueNumber): add choices in department`

File: api/api/accounts/models.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    department = models.CharField(max_length=100)
+    designation = models.CharField(max_length=100)
+    # profile_img = models.ImageField(null=True)
```
Comment (Samy-33) [inline api/api/accounts/models.py:147]: stale comment

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Comment by Samy-33** [comment]
Also, please remove empty files.

**Comment by divyanshkhetan** [comment]
@Samy-33 sir, I have made the requested changes. Please review them. Thanks!

File: api/api/accounts/managers.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from django.contrib.auth.base_user import BaseUserManager
+
+
+class CustomUserManager(BaseUserManager):
+
+    def create_user(self, email, password, **extra_fields):
+        if not email:
+            raise ValueError('The Email must be set')
+        if not password:
+            raise ValueError("User must have a password")
```
Comment (Samy-33) [inline api/api/accounts/managers.py:10]: Throwing this exception here means we are forcing users to have passwords. As we are using OAuth2 for now. That means we will not use passwords and so this check can be skipped.

File: api/api/accounts/models.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    department = models.CharField(max_length=100)
+    designation = models.CharField(max_length=100)
+
+    def __str__(self):
+        return self.user.email
+
+
+class StudentSocial(models.Model):
+    user = models.OneToOneField(get_user_model(), on_delete=models.CASCADE)
+    linkedin = models.URLField(null=True, blank=True, default=None)
```
Comment (Samy-33) [inline api/api/accounts/models.py:114]: I don't think we need to specific default in this case. If `null=True` then it'll automatically be populated as `None` by default.

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Review by Samy-33 — APPROVED** [review]
LGTM

File: api/api/accounts/models.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    objects = CustomUserManager()
+
+    def __str__(self):
+        return self.email
+
+    def save(self, *args, **kwargs):
+        """For cpanel."""
+        self.is_active = (self.is_active is True)
+        self.is_staff = (self.is_staff is True)
+        self.is_superuser = (self.is_superuser is True)
```
Comment (Samy-33) [inline api/api/accounts/models.py:72]: I don't understand this and hence don't like it. Can you remove this?

**Review by Samy-33 — CHANGES_REQUESTED** [review]

File: api/api/accounts/models.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    REQUIRED_FIELDS = []
+
+    # user manager
+    objects = CustomUserManager()
+
+    def __str__(self):
+        return self.email
+
+    def save(self, *args, **kwargs):
+        """For cpanel."""
```
Comment (Samy-33) [inline api/api/accounts/models.py:69]: Now we have this stale comment.

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Review by Samy-33 — APPROVED** [review]

**Pull request merged**